kv/txn_coord_sender.go

```diff
@@ -691,102 +717,97 @@ func (tc *TxnCoordSender) unregisterTxnLocked(txnMeta *txnMetadata) {
+	defer trace.Finalize()
 
-				request.Header().Timestamp = tc.clock.Now()
-				reply := &proto.InternalHeartbeatTxnResponse{}
-				call := proto.Call{
-					Args:  request,
-					Reply: reply,
+	// Loop with ticker for periodic heartbeats.
+	for {
+		select {
+		case <-tickChan:
+			tc.Lock()
+			proceed := true
+			txnMeta := tc.txns[id]
+			// Before we send a heartbeat, determine whether this transaction
+			// should be considered abandoned. If so, exit heartbeat.
+			if txnMeta.hasClientAbandonedCoord(tc.clock.PhysicalNow()) {
+				// TODO(tschottdorf): should we be more proactive here?
+				// The client might be continuing the transaction
+				// through another coordinator, but in the most likely
+				// case it's just gone and the open transaction record
+				// could block concurrent operations.
+				if log.V(1) {
+					log.Infof("transaction %s abandoned; stopping heartbeat",
+						txnMeta.txn)
 				}
+				proceed = false
+			}
+			// txnMeta.txn is possibly replaced concurrently,
+			// so grab a copy before unlocking.
+			txn := txnMeta.txn
+			tc.Unlock()
+			if !proceed {
+				return
+			}
 
-				tc.stopper.RunTask(func() {
-					// Each heartbeat gets its own Trace.
-					trace := tc.tracer.NewTrace(&txn)
-					ctx := tracer.ToCtx(context.Background(), trace)
-					epochEnds := trace.Epoch("heartbeat")
-					tc.wrapped.Send(ctx, call)
-					epochEnds()
-					// If the transaction is not in pending state, then we can stop
-					// the heartbeat. It's either aborted or committed, and we resolve
-					// write intents accordingly.
-					if reply.GoError() != nil {
-						log.Warningf("heartbeat to %s failed: %s", txn, reply.GoError())
-					} else if reply.Txn != nil && reply.Txn.Status != proto.PENDING {
-						tc.cleanupTxn(trace, *reply.Txn, nil)
-						proceed = false
-					}
-					trace.Finalize()
-				})
-				if !proceed {
-					return
-				}
+			request := &proto.InternalHeartbeatTxnRequest{
+				RequestHeader: proto.RequestHeader{
+					Key:  txn.Key,
+					User: security.RootUser,
+					Txn:  &txn,
+				},
+			}
 
-			case <-closer:
-				// Transaction finished.
-				return
+			request.Header().Timestamp = tc.clock.Now()
+			reply := &proto.InternalHeartbeatTxnResponse{}
+			call := proto.Call{
+				Args:  request,
+				Reply: reply,
+			}
 
-			case <-tc.stopper.ShouldStop():
-				// System shutdown.
-				return
+			epochEnds := trace.Epoch("heartbeat")
+			tc.wrapped.Send(ctx, call)
+			epochEnds()
+			// If the transaction is not in pending state, then we can stop
+			// the heartbeat. It's either aborted or committed, and we resolve
+			// write intents accordingly.
+			if reply.GoError() != nil {
+				log.Warningf("heartbeat to %s failed: %s", txn, reply.GoError())
+			} else if reply.Txn != nil && reply.Txn.Status != proto.PENDING {
+				tc.cleanupTxn(trace, *reply.Txn, nil) // signal closer
+				tickChan = nil                        // wait only for the closer, no more heartbeats
```

Comment:
no, the closer is closed on various callpaths ending up in `cleanupTxn`. This is just one of them.